Add incoherence field to Build model

JohnTortugo opened the pull request:
Relates to: https://github.com/dotnet/arcade/issues/4144
Relates to: https://github.com/dotnet/arcade/issues/4630

- Add a new field to the Maestro.Data.Models.Build class to persist information about the possible dependence incoherence of a build. The field will contain a JSON string, similarly to the Policy field in the Subscription class/table.
- Added the new field to the Maestro API.
- Patched the BuildsController to retrieve the incoherencies of a build and set the build object with the correct information.

Collateral updates:
- Updated Maestro.Client & Publish to BAR task due to existing API changes & the ones proposed in this PR.
- Had to update Maestro.DataProviders->DarcRemoteFactory to be able to create a GitClient so that I can use that RemoteFactory in Maestro.Web->BuildsController to fetch the DependencyGraph.

riarenas commented:
It seems somewhat odd to have a field named after a negative. Is the field supposed to only carry information about incoherency, or will there be something in there when a build is coherent?

JohnTortugo commented:
> It seems somewhat odd to have a field named after a negative. Is the field supposed to only carry information about incoherency, or will there be something in there when a build is coherent?

Yeah, I see what you're saying. But in this case I intend to only store incoherency information: list of incoherent properties/nodes. I think naming it "CoherencyInformation" might not be very precise... but I think would be acceptable too..

mmitche commented:
> > It seems somewhat odd to have a field named after a negative. Is the field supposed to only carry information about incoherency, or will there be something in there when a build is coherent?
> 
> Yeah, I see what you're saying. But in this case I intend to only store incoherency information: list of incoherent properties/nodes. I think naming it "CoherencyInformation" might not be very precise... but I think would be acceptable too..

Maybe just "Incoherencies" will work.

riarenas commented:
Since it's for incoherency only, I think it makes sense. I like Matt's name suggestion.

JohnTortugo commented:
I'll rename to "Incoherencies".

JohnTortugo commented:
I ended up pushing new changes to this PR instead of creating a new PR:

- I updated the PR description, PTAL.
- I restructured the `Incoherencies` field to have a similar behavior of what we have for the `Policy` field in the Subscription model.
- I added the code to retrieve the DependencyGraph to the BuildsController -> CreateBuild
- Updated the Maestro.Client sources
- Had to update some config files in Maestro.Web and Maestro.DataProviders

JohnTortugo commented:
I'm considering including the changes to BarViz (to display the incoherence infos) in this PR. Please, let me know if there is any objection.

mmitche commented on src/Maestro/Client/src/Microsoft.DotNet.Maestro.Client.csproj:
```diff
@@ -8,7 +8,7 @@
     <SwaggerOutputDirectory>$(MSBuildProjectDirectory)\Generated</SwaggerOutputDirectory>
     <SwaggerClientName>MaestroApi</SwaggerClientName>
-    <SwaggerDocumentUri>https://maestro-int.westus2.cloudapp.azure.com/api/swagger.json</SwaggerDocumentUri>
+    <SwaggerDocumentUri>https://localhost:4430/api/2020-02-20/swagger.json</SwaggerDocumentUri>
```
I think this needs to be reverted.

mmitche commented on src/Maestro/Maestro.Data/Models/Build.cs:
```diff
@@ -152,6 +153,20 @@ public string GitHubBranch
+        public BuildIncoherence BuildIncoherenciesObject
+        {
+            get => BuildIncoherenciesString == null ? null : JsonConvert.DeserializeObject<BuildIncoherence>(BuildIncoherenciesString);
+            set => BuildIncoherenciesString = value == null ? null : JsonConvert.SerializeObject(value);
```
Can this throw if non-serializable data is passed? in?

mmitche commented on src/Maestro/Maestro.Data/Models/BuildIncoherence.cs:
```diff
@@ -0,0 +1,34 @@
+        public string Commit { get; set; }
+    }
+
+    public class IncoherentNode
```
What's the difference between an incoherent node and an incoherent dependency?

JohnTortugo replied:
Talked offline. I'll remove the nodes list.

mmitche commented on src/Maestro/Maestro.Data/Models/Build.cs:
```diff
@@ -152,6 +153,20 @@ public string GitHubBranch
+        /// starting at this build.
+        /// </summary>
+        [Column("Incoherencies")]
+        public string BuildIncoherenciesString { get; set; }
```
I wonder whether we could avoid a json blob here and instead just use a table. If the interesting fields are just Name, Version, Repo, and Sha, could we just have a table with rows of Name, Version, Repo, Sha, and BuildId?

JohnTortugo replied:
Moved to a table.

JohnTortugo commented:
Pushed some changes:

- Changed the code to persist the incoherencies in a separate table
- Updated Maestro Client again
- Removed the list of incoherency nodes

mmitche commented on src/Maestro/Maestro.Data/Models/Build.cs:
```diff
@@ -151,6 +151,7 @@ public string GitHubBranch
         /// More info is available here: https://github.com/dotnet/arcade/blob/master/Documentation/CorePackages/Versioning.md#build-kind
         /// </summary>
         public bool Stable { get; set; }
+        public List<BuildIncoherence> Incoherencies { get; set; }
```
Nit: Comment for this property

mmitche commented on src/Maestro/Maestro.DataProviders/DarcRemoteFactory.cs:
```diff
@@ -2,37 +2,169 @@
+        ///     The file is assumed to be on a public endpoint.
+        ///     We return the git client executable so that this call may be easily wrapped in RetryAsync
+        /// </summary>
+        public async Task<string> EnsureLocalGit(ILogger logger)
```
What is the git executable used for here? I don't think the dependency graph paths utilize it at all.

JohnTortugo replied:
Looks like it does utilize the git executable:
- https://github.com/dotnet/arcade-services/blob/master/src/Microsoft.DotNet.Darc/src/DarcLib/Models/Darc/DependencyGraph.cs#L514
- https://github.com/dotnet/arcade-services/blob/master/src/Microsoft.DotNet.Darc/src/DarcLib/Models/Darc/DependencyGraph.cs#L846
- https://github.com/dotnet/arcade-services/blob/master/src/Microsoft.DotNet.Darc/src/DarcLib/Actions/Remote.cs#L1039
- https://github.com/dotnet/arcade-services/blob/master/src/Microsoft.DotNet.Darc/src/DarcLib/Actions/Remote.cs#L1081

mmitche replied:
It does need a git client class, but I'm pretty sure it doesnt' need git.exe. I think the get dependencies call actually just makes a GitHub or AzDO GET call.

Basically, git.exe has bitten us in the past with large numbers ending up on the box, so the fewer third party downloads we do the better.

You can probably refactor this a bit so that the places that actually need git.exe have a separate check for the availability of git.exe.

JohnTortugo replied:
Got it. Let me look look into this.

JohnTortugo replied:
Yeah, it works!

mmitche commented on src/Maestro/Maestro.Web/Api/v2020_02_20/Controllers/BuildsController.cs:
```diff
@@ -209,6 +218,15 @@ public async Task<IActionResult> Create([FromBody, Required] BuildData build)
             Data.Models.Build buildModel = build.ToDb();
             buildModel.DateProduced = DateTimeOffset.UtcNow;
+
+            if (buildModel.Incoherencies == null)
```
I would wrap this in a try/catch, logging a warning on failure. The graph build should be reliable, but:
- There are possibilities of it failing, like if github or azdo have issues.
- It doesn't directly affect the correctness of a build upload.

JohnTortugo commented:
/azp run

JohnTortugo commented:
Can you all PTAL at this again? I think now the change is 'complete' - of course pending the review.

I'll attach a screenshot of the UI changes shortly.

riarenas commented on src/Maestro/Maestro.Web/Api/v2020_02_20/Controllers/BuildsController.cs:
```diff
@@ -209,6 +218,23 @@ public async Task<IActionResult> Create([FromBody, Required] BuildData build)
+            {
+                try
+                {
+                    buildModel.Incoherencies = await 
```
isn't this an expensive operation to do on every every POST?

mmitche replied:
Yeah we need to profile this to make sure that it doesn't take too long. This doesn't look up build info any longer, so it should be much, much faster than it was before, but....
- How long does it take today?
- What happens if we hit failure cases while looking up dependencies, which could cause retries, etc. and delay this?

I can see two paths:
- We move the incoherency build to the client side, prior to creating a build
- We populate incoherencies in the queue model, like we do in some cases where we expect long operations (like triggering subscriptions), and the returned build model from Create simply does not immediately include that data

JohnTortugo replied:
To me it doesn't sound terrible expensive. I did some profiling looking at the latest ~700 builds. Here are some info:

- Most of the builds have 0 incoherencies and the computation is fast. Less than 2 seconds.
- Some builds have lost of incoherencies and computatioin take a couple dozen of seconds.

Here is a histogram excluding the coherent builds:

![time](https://user-images.githubusercontent.com/2249648/80434780-95142080-88af-11ea-88a4-51997e2810e7.png)

Do you guys think that's too slow? TBH my opinion is that the cost of a couple of seconds for the whole build isn't much.

riarenas replied:
I think that delaying the response to a post request by 10 seconds is a lot

JohnTortugo replied:
I think that for the context where the request happens, 10 seconds, or even 1 minute, isn't a lot. If we had a scenario where the user would call this API directly and wait for the response I'd fully agree that 10s is a lot. But the only scenario where this request happens is the one where responsiveness isn't a problem AFAIU.

mmitche replied:
I'm not comfortable with that length of delay. What happens if we hit a retry and end up waiting for 2 minutes? In that case, it's highly likely we'd hit an http client timeout and then the upload would fail.

That said, this field should not need to be immediately available, as it's informational only. So I think we can simply queue this operation and return the response back to the user. The incoherencies will be filled in a bit later.

To be complete, the incoherencies field could be 3 state:
- Not available (I suppose this is the case if you uploaded using an older API version?)
- Pending calculation
- Has results.

2 state might suffice too.

riarenas replied:
I like the idea of simply starting the operation without blocking the response.

Also. Thank you for getting the concrete data! Makes it so much easier to weigh the options.

mmitche replied:
>Also. Thank you for getting the concrete data! Makes it so much easier to weigh the options.

Yes, thanks for getting the data.

JohnTortugo replied:
I'll patch the code to use a `BackgroundQueue` 'event' to fill in the Incoherencies async then. Thank you all for the feedback.

mmitche commented on src/Maestro/Maestro.Web/Startup.cs:
```diff
@@ -284,17 +284,17 @@ private void ConfigureApi(IApplicationBuilder app)
-            //            a.Run(ApiRedirectHandler);
-            //        });
-            //}
+            if (HostingEnvironment.IsDevelopment() && !Program.RunningInServiceFabric())
```
Should this change be retained?

JohnTortugo replied:
I commented this block of code only to do local tests against my local BAR. I committed the code by accident.

mmitche commented:
I'm probably not the best to review the UI changes.

mmitche commented:
Awesome. Thanks for the improvements. I feel much better about this now.

JohnTortugo commented:
@alexperovich - can you PTAL on the angular part?

Here is a screenshot of how the UI will look like*:

![incoh](https://user-images.githubusercontent.com/2249648/80750452-f4a34380-8adc-11ea-81c6-25b63fca30be.png)

Ignore the repeated versions. That's a bug that was fixed in another PR.

alexperovich commented on src/Maestro/maestro-angular/src/app/page/build-graph-table/popIncoherencies.component.ts:
```diff
@@ -0,0 +1,50 @@
+  selector: 'popIncoherencies',
+  templateUrl: './popIncoherencies.component.html'
+})
+
```
NIT: There shouldn't be an empty line here. The `@Component()` thingy is a decorator, the equivalent of a C# attribute. It should be attached to the class it is applied to.

JohnTortugo replied:
Thanks. Fixed.
